### PR TITLE
[AMP] support GPU BF16 amp for dygraph

### PR DESCRIPTION
**Pull request by zhangbo9674**
### PR types
New features

### PR changes
APIs

### Describe
GPU动态图支持bf16混合精度训练。
```paddle.amp.auto_cast()```接口新增参数```dtype```，默认```float16```，可选【```float16```、```bfloat16```】
使用bfloat16进行混合精度训练示例：
```paddle.amp.auto_cast(enable=True, custom_white_list={}, custom_black_list={}, level='O2', dtype='bfloat16')```

### CODE REVIEW

File: paddle/fluid/imperative/amp_auto_cast.h [hunk 1]
```diff
@@ -33,6 +33,12 @@ enum class AmpLevel {
   O3,      // fp16
 };
 
+enum class AmpDtype {
+  D0 = 0,  // float32
+  D1,      // float16
+  D2,      // bfloat16
+};
```
Comment (zhiqiu) [inline paddle/fluid/imperative/amp_auto_cast.h]: use pten::dtype directly

File: python/paddle/fluid/tests/unittests/test_imperative_auto_mixed_precision.py [hunk 2]
```diff
@@ -861,7 +861,7 @@ def train(layer, loader, loss_fn, opt):
                           feed={feed_target_names[0]: tensor_img},
                           fetch_list=fetch_targets)
 
-        self.assertTrue(np.allclose(pred.numpy(), results, atol=1.e-5))
+        self.assertTrue(np.allclose(pred.numpy(), results, atol=1.e-2))
```
Comment (zhiqiu) [inline python/paddle/fluid/tests/unittests/test_imperative_auto_mixed_precision.py]: why change precision?

File: python/paddle/fluid/dygraph/amp/auto_cast.py [hunk 3]
```diff
@@ -125,6 +126,27 @@ def _in_pure_fp16_guard():
...
+
+
+def _is_gpu_bfloat16_supported():
+    """
+    Judge whether current gpu support bfloat16 amp.
+    """
+    prop = paddle.device.cuda.get_device_capability()
+    cuda_version = paddle.version.cuda()
+    if cuda_version is not None:
+        cuda_maj_decide = int(cuda_version.split('.')[0]) >= 11
```
Comment (zhiqiu) [inline python/paddle/fluid/dygraph/amp/auto_cast.py]: what's the meaning of 'maj_decide'?

File: python/paddle/fluid/dygraph/amp/auto_cast.py [hunk 4]
```diff
@@ -163,6 +185,7 @@ def check_optimizers(optimizers):
 @signature_safe_contextmanager
 @dygraph_only
 def amp_guard(enable=True,
+              dtype='float16',
```
Comment (zhiqiu) [inline python/paddle/fluid/dygraph/amp/auto_cast.py]: Is it better to make 'dtype' as the last parameter?

File: paddle/fluid/imperative/tracer.cc [hunk 5]
```diff
@@ -188,10 +190,16 @@ void Tracer::TraceOp(const std::string& type, const NameVarMap<VarType>& ins,
...
+      new_ins = AutoCastInputs<VarType>(type, ins);
+    }
   } else if (amp_level_ == AmpLevel::O2) {
     VLOG(5) << "Pure fp16 run operator: " << type;
-    new_ins = CastPureFp16Inputs<VarType>(type, ins);
+    if (amp_dtype_ == AmpDtype::D1) {
+      new_ins = CastPureFp16Inputs<VarType>(type, ins);
+    } else if (amp_dtype_ == AmpDtype::D2) {
+      new_ins = CastPureBf16Inputs<VarType>(type, ins);
+    }
```
Comment (zhiqiu) [inline paddle/fluid/imperative/tracer.cc]: why bf16 only supports o2?

File: python/paddle/amp/auto_cast.py [hunk 6]
```diff
@@ -19,6 +19,7 @@
 
 
 def auto_cast(enable=True,
+              dtype='float16',
```
Comment (zhiqiu) [inline python/paddle/amp/auto_cast.py]: same for the parameter

File: paddle/fluid/imperative/amp_auto_cast.h (hunk 1 above)
Comment (zhangbo9674) [reply paddle/fluid/imperative/amp_auto_cast.h]: Done, tks.

File: python/paddle/fluid/dygraph/amp/auto_cast.py (hunk 3 above)
Comment (zhangbo9674) [reply python/paddle/fluid/dygraph/amp/auto_cast.py]: Change to `cuda_version_check`.

File: python/paddle/fluid/dygraph/amp/auto_cast.py (hunk 4 above)
Comment (zhangbo9674) [reply python/paddle/fluid/dygraph/amp/auto_cast.py]: Done, tks.

File: python/paddle/amp/auto_cast.py (hunk 6 above)
Comment (zhangbo9674) [reply python/paddle/amp/auto_cast.py]: Done, tks.

File: python/paddle/fluid/tests/unittests/test_imperative_auto_mixed_precision.py (hunk 2 above)
Comment (zhangbo9674) [reply python/paddle/fluid/tests/unittests/test_imperative_auto_mixed_precision.py]: Maintain the original accuracy 1.e-5.

File: paddle/fluid/imperative/tracer.cc (hunk 5 above)
Comment (zhangbo9674) [reply paddle/fluid/imperative/tracer.cc]: supported O1：op in white list will use bf16, other will use fp32.

File: python/paddle/fluid/data_feeder.py [hunk 7]
```diff
@@ -67,6 +67,9 @@ def convert_dtype(dtype):
             # however, jointly supporting python2 and python3, (as well as python4 maybe)
             # may still be a long-lasting problem.
             return str(dtype)
+        # NOTE(zhangbo): Now numpy not support bfloat, and paddle use uint16 to carry the data of bfloat16, and there binary is consistent.
```
Comment (zhiqiu) [inline python/paddle/fluid/data_feeder.py]: ... does not support, and paddle uses unit16 to  represent ..., and there binaries are consistent.

File: python/paddle/fluid/dygraph/amp/auto_cast.py [hunk 8]
```diff
@@ -207,49 +240,91 @@ def amp_guard(enable=True,
...
+            cuda_version = paddle.version.cuda()
+            warnings.warn(
+                "For bfloat16, amp only support NVIDIA GPU with Compute Capability 8.0 or higher and CUDA Version 11.0 or higher, current GPU is: %s, with Compute Capability: %d.%d, current CUDA Version is: %s."
+                % (paddle.device.cuda.get_device_name(), prop[0], prop[1],
+                   cuda_version))
+
+    if dtype == 'float16':
+        amp_dtype = "float16"
+    elif dtype == 'bfloat16':
+        amp_dtype = "bfloat16"
```
Comment (zhiqiu) [inline python/paddle/fluid/dygraph/amp/auto_cast.py]: amp_dtype = dtype is ok

Comment (zhangbo9674) [reply python/paddle/fluid/dygraph/amp/auto_cast.py:293]: Done, tks.

File: python/paddle/fluid/data_feeder.py (hunk 7 above)
Comment (zhangbo9674) [reply python/paddle/fluid/data_feeder.py]: Done, tks.

File: python/paddle/fluid/dygraph/amp/auto_cast.py (hunk 8 above)
Comment (zhangbo9674) [reply python/paddle/fluid/dygraph/amp/auto_cast.py]: Done, tks.

**Review by zhiqiu — APPROVED** [review]
LGTM

**Review by ForFishes — APPROVED** [review]
LGTM

**Review by lanxianghit — APPROVED** [review]

**Review by TCChenlong — APPROVED** [review]
LGTM

**Pull request merged**